File: src/main/leetcode/binarysearchtreeiterator/BSTIterator.java
package binarysearchtreeiterator;

/*
*
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
*/


import java.util.*;

public class BSTIterator {
    private Stack<TreeNode> path;

    public BSTIterator(TreeNode root) {
        this.path = new Stack<>();
        TreeNode tmp = root;
        while (tmp != null) {
            this.path.push(tmp);
            tmp = tmp.left;
        }
    }

    /**
     * @return whether we have a next smallest number
     */
    public boolean hasNext() {
        return !path.isEmpty();
    }

    /**
     * @return the next smallest number
     */
    public int next() {
        TreeNode top = path.pop();

        TreeNode tmp = top.right;

        while (tmp != null) {
            path.push(tmp);
            tmp = tmp.left;
        }

        return top.val;
    }
}

File: src/main/leetcode/set0000/set100/set170/p173/BSTIterator.scala
package set0000.set100.set170.p173

/**
  * Definition for a binary tree node.
  **/
class TreeNode(var _value: Int) {
  var value: Int = _value
  var left: TreeNode = null
  var right: TreeNode = null
}

class BSTIterator(_root: TreeNode) {
  private var stacks = List.empty[TreeNode]

  var node = _root
  while (node != null) {
    stacks = node :: stacks
    node = node.left
  }

  /** @return the next smallest number */
  def next(): Int = {
    val top = stacks.head
    stacks = stacks.tail
    var node = top.right

    while (node != null) {
      stacks = node :: stacks
      node = node.left
    }

    top.value
  }

  /** @return whether we have a next smallest number */
  def hasNext(): Boolean = {
    stacks.size > 0
  }

}
